lib/trino-filesystem/src/main/java/io/trino/filesystem/encryption/EncryptionEnforcingFileSystem.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.encryption;
+
+import io.airlift.units.Duration;
+import io.trino.filesystem.FileIterator;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.TrinoOutputFile;
+import io.trino.filesystem.UriLocation;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * File system implementation that enforces encrypted file system calls.
+ */
+public class EncryptionEnforcingFileSystem
+        implements TrinoFileSystem
+{
+    private final TrinoFileSystem delegate;
+    private final EncryptionKey key;
+
+    public EncryptionEnforcingFileSystem(TrinoFileSystem delegate, EncryptionKey key)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.key = requireNonNull(key, "key is null");
+    }
+
+    @Override
+    public TrinoInputFile newInputFile(Location location)
+    {
+        return newEncryptedInputFile(location, key);
+    }
+
+    @Override
+    public TrinoInputFile newEncryptedInputFile(Location location, EncryptionKey key)
```

Comment:
It seems like these methods should either

* always fail, as we expect the non-encryption methods to be called
* fail if the passed key doesn't match the instance key

Reply:
Added a check